lib/mayaUsd/fileio/utils/meshWriteUtils.cpp

```diff
@@ -846,39 +846,46 @@ bool UsdMayaMeshWriteUtils::getMeshUVSetData(
     TfToken*       interpolation,
     VtIntArray*    assignmentIndices)
 {
-    MStatus status { MS::kSuccess };
-
-    // Sanity check first to make sure this UV set even has assigned values
-    // before we attempt to do anything with the data.
+    // Check first to make sure this UV set even has assigned values before we
+    // attempt to do anything with the data. We cannot directly use this data
+    // otherwise though since we need a uvId for every face vertex, and the
+    // returned uvIds MIntArray may be shorter than that if there are unmapped
+    // faces.
     MIntArray uvCounts, uvIds;
-    status = mesh.getAssignedUVs(uvCounts, uvIds, &uvSetName);
-    if (status != MS::kSuccess) {
-        return false;
-    }
-    if (uvCounts.length() == 0 || uvIds.length() == 0) {
+    MStatus   status = mesh.getAssignedUVs(uvCounts, uvIds, &uvSetName);
+    CHECK_MSTATUS_AND_RETURN(status, false);
+
+    if (uvCounts.length() == 0u || uvIds.length() == 0u) {
         return false;
     }
 
-    // using itFV.getUV() does not always give us the right answer, so
-    // instead, we have to use itFV.getUVIndex() and use that to index into the
-    // UV set.
+    // Transfer the UV values directly to USD, in the same order as they are in
+    // the Maya mesh.
     MFloatArray uArray;
     MFloatArray vArray;
-    mesh.getUVs(uArray, vArray, &uvSetName);
+    status = mesh.getUVs(uArray, vArray, &uvSetName);
+    CHECK_MSTATUS_AND_RETURN(status, false);
+
     if (uArray.length() != vArray.length()) {
         return false;
     }
 
-    // We'll populate the assignment indices for every face vertex, but we'll
-    // only push values into the data if the face vertex has a value. All face
-    // vertices are initially unassigned/unauthored.
-    const unsigned int numFaceVertices = mesh.numFaceVertices(&status);
     uvArray->clear();
-    assignmentIndices->assign((size_t)numFaceVertices, -1);
+    uvArray->reserve(static_cast<size_t>(uArray.length()));
+    for (unsigned int uvId = 0u; uvId < uArray.length(); ++uvId) {
+        uvArray->emplace_back(uArray[uvId], vArray[uvId]);
```

Comment:
Looks like this doesn't compile with older versions of USD. Here is an error with 20.02
```
E:\maya-usd_pr\lib\mayaUsd\fileio\utils\meshWriteUtils.cpp(876): error C2039: 'emplace_back': is not a member of 'pxrInternal_v0_20__pxrReserved__::VtArray<pxrInternal_v0_20__pxrReserved__::GfVec2f>'
```

Reply:
Ah, sorry about that. I failed to notice that `VtArray::emplace_back()`is only available from 20.11 on. Should be addressed by 6780fda93b047a6eb91ead50cbe2ca05c638bcb9.

I couldn't go as far back as 20.02, but I did verify it against 20.05 where this was also problematic.